src/lib/io/sol-memmap-storage-impl-linux.c

```diff
@@ -0,0 +1,373 @@
+            value |= (uint64_t)((uint8_t *)blob->mem)[i] << j;
+
+        ret = fread(&old_value, entry->size, 1, file);
+        if (!ret || ferror(file) || feof(file)) {
+            errno = EIO;
+            goto error;
+        }
+
+        /* We just read from file, let's rewind */
+        if (fseek(file, entry->offset, SEEK_SET) < 0)
+            goto error;
+
+        value <<= entry->bit_offset;
+        value &= mask;
+        value |= (old_value & ~mask);
+        fwrite(&value, entry->size, 1, file);
+    } else {
+        fwrite(blob->mem, sol_util_min(entry->size, blob->size), 1, file);
+    }
+
+    if (ferror(file)) {
+        errno = EIO;
+        goto error;
+    }
+
+    errno = 0;
+    if (!map_internal->file)
+        fclose(file);
+
+    if (cb)
+        cb((void *)data, name, blob, -errno);
+
+    sol_blob_unref(blob);
+
+    return -errno;
+
+error:
+    SOL_DBG("Error writing to file [%s]: %s", map_internal->resolved_path,
+        sol_util_strerrora(errno));
+    ret = -errno;
+    if (file && !map_internal->file)
+        fclose(file);
+
+    if (cb)
+        cb((void *)data, name, blob, ret);
+
+    sol_blob_unref(blob);
+
+    return ret;
+}
+
+bool
+sol_memmap_impl_perform_pending_writes(void *data)
+{
+    struct map_internal_linux *map_internal = data;
+    struct pending_write_data *pending;
+    struct sol_vector tmp_vector;
+    int i, r;
+
+    map_internal->base.timeout = NULL;
+
+    map_internal->file = fopen(map_internal->resolved_path, "r+e");
+    if (!map_internal->file) {
+        SOL_WRN("Error opening file [%s]: %s",
+            map_internal->resolved_path, sol_util_strerrora(errno));
+        return false;
+    }
+
+    tmp_vector = map_internal->base.pending_writes;
+    sol_vector_init(&map_internal->base.pending_writes,
+        sizeof(struct pending_write_data));
+
+    SOL_VECTOR_FOREACH_IDX (&tmp_vector, pending, i) {
+        sol_memmap_impl_write_raw((struct map_internal *)map_internal,
+            pending->name, pending->entry, pending->mask, pending->blob,
+            pending->cb, pending->data);
+        free(pending->name);
+        sol_blob_unref(pending->blob);
+    }
+    sol_vector_clear(&tmp_vector);
+
+    r = fclose(map_internal->file);
+    if (r == EOF)
+        SOL_WRN("Error closing file [%s]: %s",
+            map_internal->resolved_path, sol_util_strerrora(errno));
+
+    map_internal->file = NULL;
+    SOL_DBG("Performed pending writes on [%s]",
+        map_internal->resolved_path);
+
+    return false;
+}
+
+#ifdef USE_I2C
+static int
+resolve_i2c_path(const char *path, char **resolved_path)
+{
+    char *rel_path = NULL, *dev_number_s = NULL, *dev_name = NULL, *end_ptr;
+    struct sol_str_slice command = sol_str_slice_from_str(path);
+    struct sol_buffer result_path = SOL_BUFFER_INIT_EMPTY;
+    struct sol_vector instructions;
+    unsigned int dev_number;
+    int ret = -EINVAL;
+
+    instructions = sol_str_slice_split(command, ",", 5);
```

Comment:
wondering if this would be better if replaced with `sscanf()` using `%[ˆ,]` to parse `create`, `i2c`, `rel_path`...